src/utils.c

```diff
@@ -113,6 +114,65 @@ inline void never_reached(const char* fn) {
 }
 
 
+static char s3_buf[200];
+
+static SEXP s3_sym(const char* generic, const char* class) {
+  int gen_len = strlen(generic);
+  int class_len = strlen(class);
+  int dot_len = 1;
+  if (gen_len + class_len + dot_len >= 200) {
+    Rf_error("Internal error: Generic or class name is too long.");
+  }
+
+  char* buf = s3_buf;
+
+  memcpy(buf, generic, gen_len); buf += gen_len;
+  *buf = '.'; ++buf;
+  memcpy(buf, class, class_len); buf += class_len;
+  *buf = '\0';
+
+  return Rf_install(s3_buf);
+}
+
+// First check in global env, then in method table
+static SEXP get_s3_method(const char* generic, const char* class) {
+  SEXP sym = s3_sym(generic, class);
+
+  SEXP method = r_env_get(R_GlobalEnv, sym);
+  if (r_is_function(method)) {
+    return method;
+  }
+
+  method = r_env_get(vctrs_method_table, sym);
+  if (r_is_function(method)) {
+    return method;
+  }
+
+  return R_NilValue;
+}
+
+SEXP s3_method(SEXP x, const char* generic) {
```

Comment:
Need comment noting that doesn't look for default.  Maybe call it `s3_find_method()`? Also feels like `generic` should be the first arg, since methods are properties of generics.